src/include/daos_obj.h

```diff
@@ -781,7 +798,7 @@ daos_obj_fetch(daos_handle_t oh, daos_handle_t th, uint64_t flags,
  * \param[in]	th	Optional transaction handle to update with.
  *			Use DAOS_TX_NONE for an independent transaction.
  *
- * \param[in]	flags	Update flags (conditional ops).
+ * \param[in]	flags	Update flags (conditional ops or DAOS_IOF_FLATTEN).
```

Comment:
add DAOS_IOF_FLATTEN flag for daos_obj_update() because in wiki page it says -
"DAOS can provide API to allow user to indicate immutability of WORM objects:
   Application/middleware calls write(obj) with “DONE” flag."

But adding this flag to normal update request will make the obj_update() be more complex - for example request targets are different (normal update only to one dkey/group, but flatten to all groups' shards) this will make some detailed code not very convenient to handle, and possibly cause update RPC interoperability issue.

How about remove this flag's usage, only provide daos_obj_flatten() API?

Reply:
yes this would even be worse coming in from the dfs or array api where an "update" can turn into multiple daos_obj_update(). so just a single API call is better here.
that being said, the name flatten is ambiguous and has several meanings from a user perspective. there is no indication here that flatten means make immutable.
so i would suggest to rename flatten to some more informative for the user like daos_obj_set_immutable() or set_ro(). that being said, what if we want to add more properties later on the object?
maybe better to just add a new API:
daos_obj_set_prop(oid, flags);
and DAOS_OBJ_RO being one of the flags (only one for now)?

Reply:
Ok, let's only add a single API if no objection from others.

I think probably no plan to add object level property for now.
So maybe just one API - daos_obj_set_ro () is good...

Reply:
sure, until there becomes a plan to add more ;-)
so IMO better to plan ahead always for APIs rather than make API changes later.

Reply:
I agree that single API like daos_obj_set_prop(DAOS_OBJ_PROP_RO) looks better.

Reply:
daos_obj_set_prop() looks will mislead user think daos will provide object level property, I think that is not true because object level property will cause too much metadata , not we only have property for pool and container level.

Reply:
i see. anyway no strong opinion on this one. both ways sound OK to me.
i'd like to discuss though how this applies in a more generic sense, like "flattening" an entire container. we cannot expect to drive this all from the client side per object. i think better to discuss this on a meeting for now.

Reply:
in last week's md on ssd meeting, @gnailzenh suggested to start with container level flattening (flaten all objects for the container).

Reply:
@mchaarawi added container property related change for it, please check if it is fine wit you, thanks.